File: simulation/swarm/py/model_agent_test.py
import numpy as np
from absl.testing import absltest
from simulation.swarm.proto.state_pb2 import State

from simulation.swarm.py.model_agent import ModelAgent


class StaticAgentTestCase(absltest.TestCase):

    def setUp(self):
        # Configure the agent.
        agent_state = State()
        self.agent_velocity = np.array([-2, 1, 0])
        (
            agent_state.velocity.x,
            agent_state.velocity.y,
            agent_state.velocity.z,
        ) = self.agent_velocity
        self.agent = ModelAgent(agent_state)

    def test_get_principal_axes_roll(self):
        roll, pitch, yaw = self.agent.get_principal_axes()

        # Check that the roll axis is aligned to the velocity vector.
        self.assertAlmostEqual(
            np.dot(roll, self.agent_velocity),
            np.linalg.norm(roll) * np.linalg.norm(self.agent_velocity))

        # Check that the roll axis is orthogonal to the other axes.
        self.assertAlmostEqual(np.dot(roll, pitch), 0)
        self.assertAlmostEqual(np.dot(roll, yaw), 0)

    def test_get_principal_axes_pitch(self):
        roll, pitch, yaw = self.agent.get_principal_axes()

        # Check the pitch axis direction.
        expected_pitch_axis = np.array([1, 2, 0])
        self.assertAlmostEqual(
            np.dot(pitch, expected_pitch_axis),
            np.linalg.norm(pitch) * np.linalg.norm(expected_pitch_axis))

        # Check that the pitch axis is orthogonal to the other axes.
        self.assertAlmostEqual(np.dot(pitch, roll), 0)
        self.assertAlmostEqual(np.dot(pitch, yaw), 0)

    def test_get_principal_axes_yaw(self):
        roll, pitch, yaw = self.agent.get_principal_axes()

        # Check the yaw axis direction.
        expected_yaw_axis = np.array([0, 0, 1])
        self.assertAlmostEqual(
            np.dot(yaw, expected_yaw_axis),
            np.linalg.norm(yaw) * np.linalg.norm(expected_yaw_axis))

        # Check that the yaw axis is orthogonal to the other axes.
        self.assertAlmostEqual(np.dot(yaw, roll), 0)
        self.assertAlmostEqual(np.dot(yaw, pitch), 0)


if __name__ == "__main__":
    absltest.main()

File: simulation/swarm/model_agent_test.cc
#include "simulation/swarm/model_agent.h"

#include <gtest/gtest.h>

#include <Eigen/Dense>

#include "simulation/swarm/proto/state.pb.h"

namespace swarm::agent {
namespace {

// Maximum error tolerance.
constexpr double kMaxErrorTolerance = 1e-6;

class ModelAgentTest : public testing::Test {
 protected:
  // Agent velocity.
  const Eigen::Vector3d kAgentVelocity{-2, 1, 0};

  ModelAgentTest() {
    State agent_state;
    agent_state.mutable_velocity()->set_x(kAgentVelocity(0));
    agent_state.mutable_velocity()->set_y(kAgentVelocity(1));
    agent_state.mutable_velocity()->set_z(kAgentVelocity(2));
    agent_ = ModelAgent(std::move(agent_state));
  }

  // Model agent.
  ModelAgent agent_;
};

TEST_F(ModelAgentTest, GetPrincipalAxesRoll) {
  const auto principal_axes = agent_.GetPrincipalAxes();
  const Eigen::Vector3d expected_roll{-2, 1, 0};
  EXPECT_EQ(principal_axes.roll(0), expected_roll(0));
  EXPECT_EQ(principal_axes.roll(1), expected_roll(1));
  EXPECT_EQ(principal_axes.roll(2), expected_roll(2));
}

TEST_F(ModelAgentTest, GetPrincipalAxesPitch) {
  const auto principal_axes = agent_.GetPrincipalAxes();
  const Eigen::Vector3d expected_pitch{1, 2, 0};
  EXPECT_EQ(principal_axes.pitch(0), expected_pitch(0));
  EXPECT_EQ(principal_axes.pitch(1), expected_pitch(1));
  EXPECT_EQ(principal_axes.pitch(2), expected_pitch(2));
}

TEST_F(ModelAgentTest, GetPrincipalAxesYaw) {
  const auto principal_axes = agent_.GetPrincipalAxes();
  const Eigen::Vector3d expected_yaw{0, 0, 5};
  EXPECT_EQ(principal_axes.yaw(0), expected_yaw(0));
  EXPECT_EQ(principal_axes.yaw(1), expected_yaw(1));
  EXPECT_EQ(principal_axes.yaw(2), expected_yaw(2));
}

TEST_F(ModelAgentTest, GetNormalizedPrincipalAxesRoll) {
  const auto normalized_principal_axes = agent_.GetNormalizedPrincipalAxes();
  const Eigen::Vector3d expected_normalized_roll{-0.894427, 0.447214, 0};
  EXPECT_NEAR(normalized_principal_axes.roll(0), expected_normalized_roll(0),
              kMaxErrorTolerance);
  EXPECT_NEAR(normalized_principal_axes.roll(1), expected_normalized_roll(1),
              kMaxErrorTolerance);
  EXPECT_NEAR(normalized_principal_axes.roll(2), expected_normalized_roll(2),
              kMaxErrorTolerance);
}

TEST_F(ModelAgentTest, GetNormalizedPrincipalAxesPitch) {
  const auto normalized_principal_axes = agent_.GetNormalizedPrincipalAxes();
  const Eigen::Vector3d expected_normalized_pitch{0.447214, 0.894427, 0};
  EXPECT_NEAR(normalized_principal_axes.pitch(0), expected_normalized_pitch(0),
              kMaxErrorTolerance);
  EXPECT_NEAR(normalized_principal_axes.pitch(1), expected_normalized_pitch(1),
              kMaxErrorTolerance);
  EXPECT_NEAR(normalized_principal_axes.pitch(2), expected_normalized_pitch(2),
              kMaxErrorTolerance);
}

TEST_F(ModelAgentTest, GetNormalizedPrincipalAxesYaw) {
  const auto normalized_principal_axes = agent_.GetNormalizedPrincipalAxes();
  const Eigen::Vector3d expected_normalized_yaw{0, 0, 1};
  EXPECT_NEAR(normalized_principal_axes.yaw(0), expected_normalized_yaw(0),
              kMaxErrorTolerance);
  EXPECT_NEAR(normalized_principal_axes.yaw(1), expected_normalized_yaw(1),
              kMaxErrorTolerance);
  EXPECT_NEAR(normalized_principal_axes.yaw(2), expected_normalized_yaw(2),
              kMaxErrorTolerance);
}

}  // namespace
}  // namespace swarm::agent
